content/blog/unsafe-string-interning-in-go.md

```diff
@@ -0,0 +1,345 @@
++++
+date = "2020-12-29T00:00:00+00:00"
+title = "Unsafe string interning in Go"
+subtitle = "I discuss the https://go4.org/intern package and the unsafe techniques it uses to implement string interning in Go."
++++
+
+* * * * * DRAFT * * * * *
+
+For a few months, I've been collaborating on the
+[`inet.af/netaddr`](https://inet.af/netaddr) package for Go which introduces a
+new `netaddr.IP` type as an alternative to the standard library's
+[`net.IP`](https://golang.org/pkg/net/#IP) type. There are [a variety of reasons
+for this](https://pkg.go.dev/inet.af/netaddr#readme-motivation), but those are
+not the focus of today's blog.
+
+The folks over at [Tailscale](https://tailscale.com/) have been working on
+improving the performance of the package in various ways, including shrinking
+the size of types and trying out faster internal IP representations.
+Specifically, one topic that came up was [how to shrink `netaddr.IP` from 32
+bytes to 24 bytes](https://github.com/inetaf/netaddr/issues/), the same size
+as a Go slice header.
+
+Because a `netaddr.IP` can also have an associated zone for IPv6 link-local
+addresses, the type previously contained a Go string, which occupies 16 bytes.
+This is a good use case for [string
+interning](https://en.wikipedia.org/wiki/String_interning), a technique which
```

Comment:
An interesting codicil here is that when most people talk about string interning in Go, they are thinking in the context of making `[]byte` to `string` conversions cheaper by re-using the resulting `string` values. This is what github.com/josharian/intern implements. However, that implementation doesn't have the strong guarantee that identical strings have identical data pointers.

In go4.org/intern, we care a lot about the strong identity guarantee, since that's what makes them usable as map keys and for ==, yet only take one word instead of two. And there is no need to attempt to rapidly and cheaply create them out of byte slices.

It's probably worth teasing those two ideas apart easy on here to avoid confusion.

I originally wanted to add github.com/josharian/intern to go4.org/intern, but the difference is subtle enough that doing so might be dangerous/confusing.